### PR TITLE
x-pack/plugin/apm-data: add dynamic setting for enabling template registry

### LINKED ISSUE #104385: x-pack/plugin/apm-data: support dynamically enabling registry
### Description

We need to support dynamically enabling/disabling the apm-data index template registry, to support restoring a cluster from a snapshot. Basically we'll need to support what is described fro Universal Profiling here: https://www.elastic.co/guide/en/elasticsearch/reference/master/snapshots-restore-snapshot.html#restore-entire-cluster

### PR DESCRIPTION
**Pull request by axw**
Add `xpack.apm_data.registry.enabled` cluster setting, which can be used to disable the index template registry while restoring a snapshot.

Closes https://github.com/elastic/elasticsearch/issues/104385

### CODE REVIEW

**Review by danielmitterdorfer — COMMENTED** [review]
The code works fine; I left a few suggestions.

File: x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMIndexTemplateRegistry.java [hunk 1]
```diff
@@ -76,7 +79,9 @@ public APMIndexTemplateRegistry(
                 return loadIngestPipeline(pipelineConfig.getKey(), version, (List<String>) pipelineConfig.getValue().get("dependencies"));
             }).collect(Collectors.toList());
 
-            enabled = XPackSettings.APM_DATA_ENABLED.get(nodeSettings);
+            setEnabled(APMPlugin.APM_DATA_REGISTRY_ENABLED.get(nodeSettings));
+            clusterService.getClusterSettings().addSettingsUpdateConsumer(APMPlugin.APM_DATA_REGISTRY_ENABLED, this::setEnabled);
```
[the remark this answers is not in the document]
Comment (axw) [reply x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMIndexTemplateRegistry.java:83]: Good call, moved it there and introduced unit tests for APMPlugin to verify the setting behaviour.

File: x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMPlugin.java [hunk 2]
```diff
@@ -33,13 +58,19 @@ public Collection<?> createComponents(PluginServices services) {
...
         APMIndexTemplateRegistry registryInstance = registry.get();
-        logger.info("APM ingest plugin is {}", registryInstance.isEnabled() ? "enabled" : "disabled");
         registryInstance.initialize();
         return List.of(registryInstance);
     }
 
     @Override
     public void close() {
-        registry.get().close();
+        if (enabled) {
```
[the remark this answers is not in the document]
Comment (axw) [reply x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMPlugin.java]: Done, thanks.

File: x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMIndexTemplateRegistry.java [hunk 3]
```diff
@@ -106,7 +116,7 @@ protected boolean requiresMasterNode() {
 
     @Override
     protected Map<String, ComponentTemplate> getComponentTemplateConfigs() {
-        if (enabled) {
+        if (isEnabled()) {
```
[the remark this answers is not in the document]
Comment (axw) [reply x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMIndexTemplateRegistry.java]: Done, thanks.

File: x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMIndexTemplateRegistry.java [hunk 4]
```diff
@@ -86,7 +91,12 @@ public int getVersion() {
         return version;
     }
 
-    public boolean isEnabled() {
+    private synchronized void setEnabled(boolean enabled) {
```
[the remark this answers is not in the document]
Comment (axw) [reply x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMIndexTemplateRegistry.java]: Done, thanks.

File: x-pack/plugin/apm-data/src/test/java/org/elasticsearch/xpack/apmdata/APMIndexTemplateRegistryTests.java [hunk 5]
```diff
@@ -75,15 +77,23 @@ public class APMIndexTemplateRegistryTests extends ESTestCase {
...
-        clusterService = ClusterServiceUtils.createClusterService(threadPool);
+        clusterService = ClusterServiceUtils.createClusterService(threadPool, clusterSettings);
         FeatureService featureService = new FeatureService(List.of());
         stackTemplateRegistryAccessor = new StackTemplateRegistryAccessor(
             new StackTemplateRegistry(Settings.EMPTY, clusterService, threadPool, client, NamedXContentRegistry.EMPTY, featureService)
         );
+
         apmIndexTemplateRegistry = new APMIndexTemplateRegistry(
-            Settings.builder().put(APM_DATA_ENABLED.getKey(), true).build(),
+            // Settings.builder().put(APM_DATA_ENABLED.getKey(), true).put(APMPlugin.APM_DATA_REGISTRY_ENABLED.getKey(), true).build(),
```
Comment (danielmitterdorfer) [inline x-pack/plugin/apm-data/src/test/java/org/elasticsearch/xpack/apmdata/APMIndexTemplateRegistryTests.java:95]: Nit: Leftover?

**Review by danielmitterdorfer — COMMENTED** [review]
The updated code looks fine to me. `APMIndexTemplateRegistryTests#testIngestPipelines()` is failing now though; supposedly the index template registry is disabled now but the test expects an enabled one.

**Comment by axw** [comment]
Thanks @danielmitterdorfer, apparently I forgot to run all the tests after making the latest change :man_facepalming:
Now explicitly enabling the registry in the `@Before`.

File: x-pack/plugin/apm-data/src/test/java/org/elasticsearch/xpack/apmdata/APMPluginTests.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+    @After
+    @Override
+    public void tearDown() throws Exception {
+        super.tearDown();
+        apmPlugin.close();
+        threadPool.shutdownNow();
+    }
+
+    public void testRegistryEnabledSetting() throws Exception {
```
Comment (danielmitterdorfer) [inline x-pack/plugin/apm-data/src/test/java/org/elasticsearch/xpack/apmdata/APMPluginTests.java:55]: Just as a suggestion: If you want to be extra cautious you could also add a test to ensure the registry is disabled if the plugin is disabled.

**Review by danielmitterdorfer — APPROVED** [review]
LGTM. I left one optional suggestion.

**Comment by axw** [comment]
@elasticmachine update branch

File: x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMPlugin.java [hunk 7]
```diff
@@ -10,36 +10,62 @@
...
+            new APMIndexTemplateRegistry(settings, clusterService, services.threadPool(), services.client(), services.xContentRegistry())
         );
-        APMIndexTemplateRegistry registryInstance = registry.get();
-        logger.info("APM ingest plugin is {}", registryInstance.isEnabled() ? "enabled" : "disabled");
-        registryInstance.initialize();
-        return List.of(registryInstance);
+        if (enabled) {
+            APMIndexTemplateRegistry registryInstance = registry.get();
+            registryInstance.setEnabled(APM_DATA_REGISTRY_ENABLED.get(settings));
+            clusterService.getClusterSettings().addSettingsUpdateConsumer(APM_DATA_REGISTRY_ENABLED, registryInstance::setEnabled);
```
Comment (gmarouli) [inline x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMPlugin.java:56]: We tend to encapsulate this in the code of the class using the `enabled` flag, in this case the `APMIndexTemplateRegistry`. We usually read the value of the setting and we set it in the constructor and in a method like `initialize()` we register the listener. This way it's transparent in the `APMIndexTemplateRegistry` how the enabled flag is changing.

Comment (axw) [reply x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMPlugin.java:56]: @gmarouli is it worth me changing now? I can go back and update it.

The only way the `enabled` flag is updated is through the `setEnabled` method, and I'm not sure why `APMIndexTemplateRegistry` would care what caused that method to be called -- but if it's conventional and this difference will be confusing, I can go back and update it.

Comment (gmarouli) [reply x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMPlugin.java:56]: It's more about encapsulation, the logic of how the flag for this class is changing is contained in the class itself. It's ok for now, if we come back to this we can address it.

**Pull request merged**